src/wp-admin/includes/file.php

```diff
@@ -1934,12 +1934,18 @@ function copy_dir( $from, $to, $skip_list = array() ) {
 function move_dir( $from, $to ) {
```

Comment:
Rather than conditionally checking the type of file system been used here, I think it would be better to create `WP_Filesystem_*::move_dir()`.

For anything that supports renaming of directories, it might be able to be an alias for the existing `WP_Filesystem_*::move()` function. On my reading of the docs, `rename()`, `ftp_rename()` supports it. `ssh2_sftp_rename()` is documented as file only. I am unclear about ftp sockets.

For consistency with the existing rename functions, it would be good to do.

```suggestion
function move_dir( $from, $to, $overwrite = false ) {
```

Reply:
The function that I'm trying to fix is `copy_dir()`. It's not a `WP_Filesystem` function.

Reply:
The reason for changing the `$wp_filesystem->move()` to `move_dir()` in the backup functions is that if the `$wp_filesytem` function degrades from a directory move to a single file copy.

Reply:
I'm inclined to agree at first, that calling `rename()` isn't a problem here - it doesn't matter if a move operation happens via the FS layer or via PHP IO, a move operation will retain ownership/permissions/etc. As long as the files were created via the FS layer (or existed prior to the upgrade routine running).

However, PHP's `rename()` isn't always a `move` operation, it's sometimes a `copy` operation, where the source and destination exist on different filesystem mounts. This is partially resolved by WordPress only using `wp-content/upgrade` as the temporary directory, and it's unlikely that `wp-content/plugins` will exist on a different filesystem - but if you step back and look at core upgrades that's more likely although still unlikely.

So I'm not 100% sure here, It should be fine, but there might be some edge-cases where it's not, but I'm not sure if that's something to be concerned about.

https://www.php.net/manual/en/function.rename.php#117590
> More explicitly, rename() may still return (bool) true, despite the warnings that result from the underlying calls to chown() or chmod(). This behavior can be misleading absent a deeper understanding of the underlying mechanics. To rename across filesystems, PHP "fakes it" by calling copy(), unlink(), chown(), and chmod() (not necessarily in that order). See PHP bug #50676 for more information.